Pulldasher: Add the 'mergeable' field

jordycosta opened the pull request:
### Issue overview:

1. Pull requests with merge conflicts can end up in the **Ready** column in Pulldasher
2. QA is unable to pull up the latest changes to a pull request if there are conflicts present

![example](https://user-images.githubusercontent.com/95656772/224138987-91aa5413-0413-4ff7-92d5-dc31c46a81d8.png)

### Resolution:

* Send pull requests that have fulfilled CR/QA requirements _(but have a merge conflict)_, to the **Deploy Blocked** column.
* Update QA column sorting to place pull requests without merge conflicts above those with merge conflicts.
* Add a new Pulldasher flag to easily discern if a pull request has merge conflicts.

---

### Changes Made:

1. Added in a new field from GitHub _(mergeable)_
2. Create a new flag for the new field
3. Updated Pulldasher **QA** column sorting logic 
4. Updated Pulldasher **Ready** + **Deploy Blocked** column pull logic

>Documentation reference:
https://docs.github.com/en/rest/pulls/pulls?apiVersion=2022-11-28#get-a-pull-request

---

### CR/QA Notes:

CR: Review to https://github.com/iFixit/pulldasher/pull/356/commits/88db70109f0aa106040bd2afe84d455b2bd26df4

QA:

1. Checkout | merge with [`add-mergeable-field`](https://github.com/iFixit/pulldasher/tree/add-mergeable-field)
2. `update-pulldasher-dev --run .`
3. Navigate to https://pulldasher-dev.cominor.com/
4. Ensure that pull requests with merge conflicts:

    a.) Have a flag applied
    b.) Are in the **Deploy Blocked** column if requirements are fulfilled
    c.) Are sorted lower in the **QA** column

---

Closes #351

danielbeardsley commented on frontend/src/pulldasher/index.tsx:
```diff
@@ -33,9 +33,9 @@ function Pulldasher() {
   const pulls = useAllOpenPulls();
   const pullsCIBlocked = pulls.filter((pull) => pull.isCiBlocked());
-  const pullsDeployBlocked = pulls.filter((pull) => pull.isDeployBlocked());
+  const pullsDeployBlocked = pulls.filter((pull) => pull.isDeployBlocked() || pull.hasMergeConflicts);
```
```suggestion
  const pullsDeployBlocked = pulls.filter((pull) => pull.isDeployBlocked() || pull.hasMergeConflicts());
```

jordycosta replied:
Good catch, thank you!

danielbeardsley commented on frontend/src/pulldasher/index.tsx:
```diff
@@ -33,9 +33,9 @@ function Pulldasher() {
+  const pullsDeployBlocked = pulls.filter((pull) => pull.isDeployBlocked() || pull.hasMergeConflicts);
   const pullsReady = pulls.filter(
-    (pull) => pull.isReady() && pull.isCiRequired()
+    (pull) => pull.isReady() && pull.isCiRequired() && !pull.hasMergeConflicts()
```
I might augment the `isReady()` function to check for merge conflicts instead of adding this each place `isReady()` is used.

danielbeardsley commented on frontend/src/pull-card/flags.tsx:
```diff
@@ -86,6 +88,13 @@ export const Flags = memo(function Flags({ pull }: { pull: Pull }) {
           icon={faSnowflake}
         />
       )}
+      {mergeConflict && (
```
I think I'd use `pull.hasMergeConflicts()` directly here to reduce the indirection.

jordycosta replied:
I agree, resolved via https://github.com/iFixit/pulldasher/pull/356/commits/88db70109f0aa106040bd2afe84d455b2bd26df4

danielbeardsley commented:
Looking good!

danielbeardsley commented:
CR :+1: 

Great!

danielbeardsley reviewed (approved):
CR :+1: 

Good job!

danielbeardsley commented:
FYI, this is what the flag looks like:
![image](https://user-images.githubusercontent.com/26855/224116210-d1c649ce-4049-4d8f-8500-14a568e9f056.png)


Oh wait.. dev_block :+1:  on adding an example to the https://pulldasher-dev.cominor.com/pull-card-demo.html

danielbeardsley reviewed (approved):
CR :+1:

jordycosta commented:
un_dev_block 🟢 

Added an example for the `draft` & `mergeable` fields to /pull-card-demo.html on Pulldasher-Dev

danielbeardsley reviewed (approved):
CR :+1:

jordycosta commented:
dev_block 🟥 on updating Deploy Blocked column logic

jordycosta commented:
un_dev_block 🟢

danielbeardsley reviewed (dismissed):
CR :+1: 

Nice!

jordycosta commented:
Switched the flag color and added some sorting for the **Deploy Blocked** column via https://github.com/iFixit/pulldasher/pull/356/commits/479ce5d0c376db5903f26b11b1551bee2c0d5b9f

danielbeardsley reviewed (approved):
CR :+1:

jordycosta commented:
QA 🟢 

I think that this is good to go 👍 

![Screenshot 2023-03-10 093348](https://user-images.githubusercontent.com/95656772/224384072-7967e111-708e-4949-a453-e7bd8b9400bc.jpg)

---

Confirmed:

1. Pull requests with merge conflicts have a flag shown

> Since there are a couple of state specific specs to this pull request that are more difficult to reproduce, I opt to monitor for quality assurance overtime and make modifications if necessary

erinemay commented:
@jordycosta, I noticed a pull on Pulldasher that doesn't show the merge conflicts indicator. 

https://github.com/iFixit/react-commerce/pull/1439
![image](https://user-images.githubusercontent.com/72166715/225136692-caba88ae-250c-4433-a7c7-efd7509dd81b.png)
![image](https://user-images.githubusercontent.com/72166715/225136853-4205e94a-673b-4112-a463-80df2d8665ca.png)

deltuh-vee commented:
@erinemay I don't think that pull has a merge conflict. It's just out of date with main and no one has done an approving CR on it.
